packages/grafana-prometheus/src/components/monaco-query-field/monaco-completion-provider/data-provider.ts

```diff
@@ -0,0 +1,117 @@
+import { HistoryItem } from '@grafana/data';
+// @ts-ignore
+import type { Monaco } from '@grafana/ui';
+
+import PromQlLanguageProvider from '../../../language_provider';
+import { PromQuery } from '../../../types';
+
+export const CODE_MODE_SUGGESTIONS_INCOMPLETE_EVENT = 'codeModeSuggestionsIncomplete';
+
+export type SuggestionsIncompleteEvent = CustomEvent<{
+  limit: number;
+  datasourceUid: string;
+}>;
+
+export function isSuggestionsIncompleteEvent(e: Event): e is SuggestionsIncompleteEvent {
+  return (
+    e.type === CODE_MODE_SUGGESTIONS_INCOMPLETE_EVENT &&
+    'detail' in e &&
+    typeof e.detail === 'object' &&
+    e.detail !== null &&
+    'limit' in e.detail &&
+    'datasourceUid' in e.detail
+  );
+}
+
+interface Metric {
+  name: string;
+  help: string;
+  type: string;
+}
+
+export interface DataProviderParams {
+  languageProvider: PromQlLanguageProvider;
+  historyProvider: Array<HistoryItem<PromQuery>>;
+}
+
+export class DataProvider {
+  readonly languageProvider: PromQlLanguageProvider;
+  readonly historyProvider: Array<HistoryItem<PromQuery>>;
+  readonly getSeriesLabels: typeof this.languageProvider.getSeriesLabels;
+  readonly getSeriesValues: typeof this.languageProvider.getSeriesValues;
+  readonly getAllLabelNames: typeof this.languageProvider.getLabelKeys;
+  readonly getLabelValues: typeof this.languageProvider.getLabelValues;
+  readonly metricNamesSuggestionLimit: number;
+  /**
+   * The text that's been typed so far within the current {@link Monaco.Range | Range}.
+   *
+   * @remarks
+   * This is useful with fuzzy searching items to provide as Monaco autocomplete suggestions.
+   */
+  private inputInRange: string;
+  private suggestionsIncomplete: boolean;
+
+  constructor(params: DataProviderParams) {
+    this.languageProvider = params.languageProvider;
+    this.historyProvider = params.historyProvider;
+    this.inputInRange = '';
+    this.metricNamesSuggestionLimit = this.languageProvider.datasource.metricNamesAutocompleteSuggestionLimit;
+    this.suggestionsIncomplete = false;
+    this.getSeriesLabels = this.languageProvider.getSeriesLabels.bind(this.languageProvider);
+    this.getSeriesValues = this.languageProvider.getSeriesValues.bind(this.languageProvider);
+    this.getAllLabelNames = this.languageProvider.getLabelKeys.bind(this.languageProvider);
+    this.getLabelValues = this.languageProvider.getLabelValues.bind(this.languageProvider);
+  }
+
+  getHistory(): string[] {
+    return this.historyProvider.map((h) => h.query.expr).filter((expr) => expr !== undefined);
```

Comment:
We don't check against `null`. Do you think we should use `filter(Boolean)`?

Additionally, can't we do filtering without mapping?: 
```
return this.historyProvider.filter((h) => h.query.expr !== undefined);
```

Reply:
> Do you think we should use filter(Boolean)?

Totally! Updated in 91c272c3fd0e025af3904c2a13ba38e9afe439b1.

> Additionally, can't we do filtering without mapping?
> ```
> return this.historyProvider.filter((h) => h.query.expr !== undefined);
> ```

We still need to map to get the `query.expr` from each history item (return type should be `string[]` and not `HistoryItem<PromQuery>[]`).